File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        return isSubTreePresent(root, subRoot);
    }

    private boolean isSubTreePresent(TreeNode root, TreeNode subRoot) {
        // base condition
        if (root == null) {
            return false;
        }
        if (checkEqual(root, subRoot)) {
            return true;
        }

        boolean left = isSubTreePresent(root.left, subRoot);
        boolean right = isSubTreePresent(root.right, subRoot);

        return left || right;
    }

    private boolean checkEqual(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) {
            return true;
        }
        if (root == null || subRoot == null) {
            return false;
        }
        if (root.val != subRoot.val) {
            return false;
        }
        
        return checkEqual(root.left, subRoot.left) && checkEqual(root.right, subRoot.right);
    }
}

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {

        if (root == nullptr)
            return nullptr;

        if (key < root->val) {
            root->left = deleteNode(root->left, key);
        }
        else if (key > root->val) {
            root->right = deleteNode(root->right, key);
        }
        else {

            // 0 children OR only right child
            if (root->left == nullptr)
                return root->right;

            // only left child
            if (root->right == nullptr)
                return root->left;

            // 2 children
            TreeNode* successor = root->right;

            while (successor->left != nullptr) {
                successor = successor->left;
            }

            root->val = successor->val;

            root->right = deleteNode(root->right, successor->val);
        }

        return root;
    }
};